### PR TITLE
Update pipeline state on server as a whole on approval

### LINKED ISSUE #3472: Approving pipeline lists workflows twice
### Component

web-ui

### Describe the bug

Approving a pipeline lists its workflows twice. Duplicates are removed as soon as the pipeline starts.

probably a regression of #2757 

### System Info

```shell
ci.woodpecker-ci.org
```

### Additional context

grafik

### Validations

- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't [already an issue](https://github.com/woodpecker-ci/woodpecker/issues) that reports the same bug to avoid creating a duplicate.
- [X] Checked that the bug isn't fixed in the `next` version already [https://woodpecker-ci.org/faq#which-version-of-woodpecker-should-i-use]

### LINKED ISSUE #3494: Duplicate step/service names results in conflicting container names
### Component

server

### Describe the bug

If a workflow contains multiple steps or services with the same name (one service and one step with the same name is affected as well) it will result in a race condition during the container creation.

image

image
image
 

### System Info

```shell
next-5d98a717ac
```

### Additional context

caused by https://github.com/woodpecker-ci/woodpecker/pull/2757/files#diff-6f809c0a6ea3b1924d8664a9f0dba9e308c15a5915d6444a18f0cb913809e26aR90

### Validations

- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't [already an issue](https://github.com/woodpecker-ci/woodpecker/issues) that reports the same bug to avoid creating a duplicate.
- [X] Checked that the bug isn't fixed in the `next` version already [https://woodpecker-ci.org/faq#which-version-of-woodpecker-should-i-use]

### PR DESCRIPTION
**Pull request by 6543**
We can not just update some records for steps, as we want the pipeline engine as single source of truth but not manage the state.
And the server should only manage the state but not how pipelines work.

We can match the pipeline but neither workflows or steps 1:1, so we "update" them as a whole by deleting existing workflow and step data and insert the new info from engine.

close   #3494
close  #3472

---
*Sponsored by Kithara Software GmbH*

### CODE REVIEW

File: server/pipeline/approve.go [hunk 1]
```diff
@@ -79,24 +79,15 @@ func Approve(ctx context.Context, store store.Store, currentPipeline *model.Pipe
...
-								step.UUID = storeStep.UUID
-								break
-							}
-						}
-					}
-				}
-
-				break
-			}
+	// we have now way to link old workflows & steps in database to new engine generated steps.
```
Comment (xoxys) [inline server/pipeline/approve.go:82]:
```suggestion
	// we have no way to link old workflows and steps in database to new engine generated steps,
```

**Comment by 6543** [comment]
I updated the pull description why we do things exactly that way ...

**Comment by xoxys** [comment]
Is the delete and write done in a single transaction?

**Comment by 6543** [comment]
No ... I'll address it 1mom

File: server/pipeline/items.go [hunk 2]
```diff
@@ -125,6 +125,9 @@ func setPipelineStepsOnPipeline(pipeline *model.Pipeline, pipelineItems []*stepb
 		}
 	}
 
+	// the workflows in the pipeline should be empty as only we do populate them,
+	// but it a pipeline was already loaded form database it might contain things, so we just clean it
```
Comment (6543) [inline server/pipeline/items.go]:
```suggestion
	// but if a pipeline was already loaded form database it might contain things, so we just clean it
```

**Comment by 6543** [comment]
finding a better title is welcome ;)

File: server/store/store.go [hunk 3]
```diff
@@ -171,7 +171,7 @@ type Store interface {
 	// Workflow
 	WorkflowGetTree(*model.Pipeline) ([]*model.Workflow, error)
 	WorkflowsCreate([]*model.Workflow) error
-	WorkflowsDelete(*model.Pipeline) error
+	WorkflowsSwitch(*model.Pipeline, []*model.Workflow) error
```
Comment (xoxys) [inline server/store/store.go:174]:
```suggestion
	WorkflowsReplace(*model.Pipeline, []*model.Workflow) error
```

File: server/pipeline/approve.go [hunk 4]
```diff
@@ -81,14 +81,9 @@ func Approve(ctx context.Context, store store.Store, currentPipeline *model.Pipe
...
 	// so we just delete the old and insert the new ones
-	{
-		if err := store.WorkflowsDelete(currentPipeline); err != nil {
-			return nil, err
-		}
-		if err := store.WorkflowsCreate(currentPipeline.Workflows); err != nil {
-			log.Error().Err(err).Str("repo", repo.FullName).Msgf("error persisting steps for %s#%d", repo.FullName, currentPipeline.Number)
-			return nil, err
-		}
+	if err := store.WorkflowsSwitch(currentPipeline, currentPipeline.Workflows); err != nil {
```
Comment (xoxys) [inline server/pipeline/approve.go:84]:
```suggestion
	if err := store.WorkflowsReplace(currentPipeline, currentPipeline.Workflows); err != nil {
```

File: server/store/datastore/workflow.go [hunk 5]
```diff
@@ -53,12 +61,26 @@ func (s storage) WorkflowsCreate(workflows []*model.Workflow) error {
...
 	}
-
-	return sess.Commit()
+	return nil
 }
 
-func (s storage) WorkflowsDelete(pipeline *model.Pipeline) error {
-	return s.workflowsDelete(s.engine.NewSession(), pipeline.ID)
+// WorkflowsSwitch do "update" workflows and related steps by deleting all existing workflow and steps and insert the new ones
+func (s storage) WorkflowsSwitch(pipeline *model.Pipeline, workflows []*model.Workflow) error {
```
Comment (xoxys) [inline server/store/datastore/workflow.go:68]:
```suggestion
// WorkflowsReplace performs an atomic replacement of workflows and associated steps by deleting all existing workflows and steps and inserting the new ones
func (s storage) WorkflowsReplace(pipeline *model.Pipeline, workflows []*model.Workflow) error {
```

File: server/store/mocks/store.go [hunk 6]
```diff
@@ -2609,6 +2419,24 @@ func (_m *Store) WorkflowsCreate(_a0 []*model.Workflow) error {
 	return r0
 }
 
+// WorkflowsSwitch provides a mock function with given fields: _a0, _a1
+func (_m *Store) WorkflowsSwitch(_a0 *model.Pipeline, _a1 []*model.Workflow) error {
```
Comment (6543) [inline server/store/mocks/store.go:2423]:
```suggestion
// WorkflowsReplace provides a mock function with given fields: _a0, _a1
func (_m *Store) WorkflowsReplace(_a0 *model.Pipeline, _a1 []*model.Workflow) error {
```

File: server/store/mocks/store.go [hunk 7]
```diff
@@ -2609,6 +2419,24 @@ func (_m *Store) WorkflowsCreate(_a0 []*model.Workflow) error {
...
+	ret := _m.Called(_a0, _a1)
+
+	if len(ret) == 0 {
+		panic("no return value specified for WorkflowsSwitch")
```
Comment (6543) [inline server/store/mocks/store.go:2427]:
```suggestion
		panic("no return value specified for WorkflowsReplace")
```

**Review by xoxys — APPROVED** [review]
LGTM, tested locally.

**Pull request merged**